### ISSUE TITLE
descending specificity with direct child combinator

### ISSUE THREAD

**Issue by jacekkopecky** [stylelint/stylelint#3072]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A bug where the rule triggers incorrectly.

> Which rule, if any, is this issue related to?

`no-descending-specificity`

> What CSS is needed to reproduce this issue?

```css
body > header > h1 {
  width: 10em;
}

body > h1 {
  width: 20em;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "no-descending-specificity": true
  }
}
```

> Which version of stylelint are you using?

`8.4.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint file.css` and in Atom"

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No non-standard syntax.

> What did you expect to happen?

No warnings to be flagged, because the two rules never match the same h1.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

```shell
file.css
 5:1  ✖  Expected selector "body > h1" to come before selector "body >    no-descending-specificity
         header > h1"
```

**Comment by hudochenkov** [comment]
Thanks for the report and for using the template.

The rule works as expected. It checks [specificity](https://www.w3.org/TR/css3-selectors/#specificity) only.

`body > header > h1` has specificity 3.
`body > h1` has specificity 2.
`body h1` has specificity 2.

**Comment by jacekkopecky** [comment]
Hi, yes, in that case this would be more a feature request than a bug, is that correct?

The `no-descending-specificity` rule is meant to prevent the author thinking they are overriding some rules with later rules when they actually aren't. Checking that this situation doesn't apply between `body > header > h1` and `body > h1` is extra work for style-lint. Do you think it would be difficult to do @hudochenkov ?

**Comment by jeddy3** [comment]
> rule is meant to prevent the author thinking they are overriding some rules with later rules when they actually aren't.

I think the original purpose of the rule is more general than that. It's used to avoid a, potentially confusing, clash of two mechanisms for prioritization. From the [README](https://github.com/stylelint/stylelint/tree/master/lib/rules/no-descending-specificity):

> Disallow selectors of lower specificity from coming after overriding selectors of higher specificity.

> Source order is important in CSS, and when two selectors have the *same* specificity, the one that occurs *last* will take priority. However, the situation is different when one of the selectors has a *higher* specificity. In that case, source order does *not* matter: the selector with higher specificity will win out even if it comes first.

> The clashes of these two mechanisms for prioritization, source order and specificity, can cause some confusion when reading stylesheets. If a selector with higher specificity comes *before* the selector it overrides, we have to think harder to understand it, because it violates the source order expectation. **Stylesheets are most legible when overriding selectors always come *after* the selectors they override.** That way both mechanisms, source order and specificity, work together nicely.

In the example above:

```css
body > header > h1 {
  width: 10em;
}

body > h1 {
  width: 20em;
}
```

The source order and the order of specificity do clash.

> Hi, yes, in that case this would be more a feature request than a bug, is that correct?

Having said all that, the change proposed in https://github.com/stylelint/stylelint/issues/2489 might cast some light on whether there's value in adding an option to allow this behaviour. @jacekkopecky Would you like to consider [contributing the fix](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#fixing-a-bug-in-an-existing-rule) for https://github.com/stylelint/stylelint/issues/2489? We can see where to go after that.
